qcog_python_client/qcog/pytorch/agent.py

```diff
@@ -0,0 +1,140 @@
+"""Class to handle all the PyTorch operations."""
+
+from typing import Any, Callable, Coroutine, cast
+
+from qcog_python_client.qcog.pytorch.discover._discover import (
+    DiscoverCommand,
+    DiscoverHandler,
+)
+from qcog_python_client.qcog.pytorch.handler import Handler, ToolFn
+from qcog_python_client.qcog.pytorch.parameters._saveparameters import (
+    SaveParametersCommand,
+    SaveParametersHandler,
+)
+from qcog_python_client.qcog.pytorch.train._train import TrainCommand, TrainHandler
+from qcog_python_client.qcog.pytorch.upload._upload import UploadHandler
+from qcog_python_client.qcog.pytorch.validate._validate import ValidateHandler
+
+
+class PyTorchAgent:
+    """PyTorch Agent."""
+
+    def __init__(self) -> None:
+        """Initialize the PyTorch Agent."""
+        # Tools should be async functions that return a dictionary
+        self.tools: dict[str, Callable[..., Coroutine[Any, Any, dict]]] = {}
+        self._chain: Handler | None = None
+
+    @property
+    def chain(self) -> Handler:
+        """Get the chain of responsibility."""
+        if not self._chain:
+            raise ValueError("Chain not initialized. Call the `init` method first.")
+        return self._chain
+
+    def _get_tool(self, tool_name: str) -> ToolFn:
+        """Get a tool from the tools dictionary."""
+        tool = self.tools.get(tool_name)
+
+        if not tool:
+            raise ValueError(
+                f"Tool {tool_name} not found. Available tools: {list(self.tools.keys())}"  # noqa
+            )
+
+        return tool
+
+    def set_context(self, key: str, value: Any) -> None:
+        """Set a value on the chain context."""
+        self.chain.context[key] = value
+
+    def init(self, *chain: Handler) -> Handler:
+        """Initialize the Responsibility Chain."""
+        if len(chain) == 1:
```

Comment:
`len(chain) < 2` I think this code will succeed with 0 args and then freak out later in a weird place

Reply:
another shame